File: 12/12.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const START = 0
const END = 1000

type Path struct{ start, end string }
type CaveList []int
type CaveMap map[int]CaveList
type PathList []int
type StackType PathList
type Stack []StackType
type Rejecter func(int, []int) bool

func (stack *Stack) push(b StackType) { *stack = append(*stack, b) }
func (stack *Stack) peek() StackType  { return (*stack)[len(*stack)-1] }
func (stack *Stack) isEmpty() bool    { return len(*stack) == 0 }
func (stack *Stack) pop() StackType {
	last := len(*stack) - 1
	rval := (*stack)[last]
	*stack = (*stack)[:last]
	return rval
}

func createPath(line string) Path {
	split := strings.Split(line, "-")
	return Path{split[0], split[1]}
}

func isSmall(cave string) bool {
	return cave[0] >= 'a' && cave[0] <= 'z'
}

func weight(cave string) int {
	if isSmall(cave) {
		return -1
	}
	return 1
}

func readFile(name string) CaveMap {
	file, _ := os.Open(name)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var caveMap = make(CaveMap)
	var names = make(map[string]int)
	names["start"] = START
	names["end"] = END
	next := 1
	for scanner.Scan() {
		path := createPath(scanner.Text())
		if _, exists := names[path.start]; !exists {
			names[path.start] = next * weight(path.start)
			next++
		}
		if _, exists := names[path.end]; !exists {
			names[path.end] = next * weight(path.end)
			next++
		}
		caveMap[names[path.start]] = append(caveMap[names[path.start]], names[path.end])
		caveMap[names[path.end]] = append(caveMap[names[path.end]], names[path.start])
	}
	return caveMap
}

func partOne(next int, work []int) bool {
	if next > 0 {
		return false
	}
	for _, w := range work {
		if w == next {
			return true
		}
	}
	return false
}

func partTwo(next int, work []int) bool {
	if next == START {
		return true
	}
	if !partOne(next, work) {
		return false
	}
	for i := 1; i < len(work); i++ {
		if work[i] > 0 {
			continue
		}
		for j := i + 1; j < len(work); j++ {
			if work[i] == work[j] {
				return true
			}
		}
	}
	return false
}

func pathCount(caveMap CaveMap, rejecter Rejecter) int {
	var workingPaths = make(Stack, 0)
	rval := 0
	workingPaths.push([]int{START})
	for !workingPaths.isEmpty() {
		work := workingPaths.pop()
		for _, next := range caveMap[work[len(work)-1]] {
			if rejecter(next, work) {
				continue
			}
			if next == END {
				rval++
			} else {
				prospective := make([]int, len(work))
				copy(prospective, work)
				prospective = append(prospective, next)
				workingPaths.push(prospective)
			}
		}
	}
	return rval
}

func main() {
	caveMap := readFile("input.test")
	fmt.Println(pathCount(caveMap, partOne)) // 10
	fmt.Println(pathCount(caveMap, partTwo)) // 36
	caveMap = readFile("input.test2")
	fmt.Println(pathCount(caveMap, partOne)) // 19
	fmt.Println(pathCount(caveMap, partTwo)) // 103
	caveMap = readFile("input.test3")
	fmt.Println(pathCount(caveMap, partOne)) // 226
	fmt.Println(pathCount(caveMap, partTwo)) // 3509
	caveMap = readFile("input")
	fmt.Println(pathCount(caveMap, partOne)) // 4304
	fmt.Println(pathCount(caveMap, partTwo)) // 118242
}

File: 12/12.cc
#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <vector>

namespace {

constexpr auto START = 0;
constexpr auto END = 1000;
struct Path { std::string start, end; };
using CaveList = std::vector<int>;
using CaveMap = std::map<int, CaveList>;
using PathList = std::vector<int>;

auto &operator>>(std::istream &in, Path &path) { // "createPath"
  auto line = std::string{};
  in >> line;
  path.start = line.substr(0, line.find('-'));
  path.end = line.substr(line.find('-') + 1);
  return in;
}

auto is_small(const std::string &cave) {
  return cave[0] >= 'a' && cave[0] <= 'z';
}

auto weight(const std::string &cave) { return is_small(cave) ? -1 : 1; }

auto read_file(const std::string &name) {
  auto file = std::ifstream{name};

  auto caveMap = CaveMap{};
  auto names = std::map<std::string, int>{};
  names["start"] = START;
  names["end"] = END;
  auto next = 1;
  auto path = Path{};
  while (file >> path) {
    if (names.find(path.start) == names.end()) {
      names[path.start] = weight(path.start) * next++;
    }
    if (names.find(path.end) == names.end()) {
      names[path.end] = weight(path.end) * next++;
    }
    caveMap[names[path.start]].push_back(names[path.end]);
    caveMap[names[path.end]].push_back(names[path.start]);
  }
  return caveMap;
}

auto partOne(int next, const PathList &work) {
  return next <= 0 && std::find(work.begin(), work.end(), next) != work.end();
}

auto partTwo(int next, const PathList &work) {
  if (next == 0) {
    return true;
  }
  if (!partOne(next, work)) {
    return false;
  }
  for (auto i = 1; i < work.size(); ++i) {
    if (work[i] > 0) {
      continue;
    }
    for (auto j = i + 1; j < work.size(); ++j) {
      if (work[i] == work[j]) {
        return true;
      }
    }
  }
  return false;
}

template <class Rejecter>
auto pathCount(const CaveMap &caveMap, Rejecter rejecter) {
  auto workingPaths = std::vector<PathList>{};
  auto rval = 0;
  workingPaths.push_back(PathList{START});
  while (workingPaths.size()) {
    auto work = workingPaths.back();
    workingPaths.pop_back();
    for (auto &next : caveMap.at(work.back())) {
      if (rejecter(next, work)) {
        continue;
      }
      if (next == END) {
        ++rval;
      } else {
        auto prospective = work;
        prospective.push_back(next);
        workingPaths.push_back(prospective);
      }
    }
  }
  return rval;
}

} // namespace

auto main() -> int {
  auto caveMap = read_file("input.test");
  std::cout << pathCount(caveMap, partOne) << '\n'  // 10
            << pathCount(caveMap, partTwo) << '\n'; // 36
  caveMap = read_file("input.test2");
  std::cout << pathCount(caveMap, partOne) << '\n'  // 19
            << pathCount(caveMap, partTwo) << '\n'; // 103
  caveMap = read_file("input.test3");
  std::cout << pathCount(caveMap, partOne) << '\n'  // 226
            << pathCount(caveMap, partTwo) << '\n'; // 3509
  caveMap = read_file("input");
  std::cout << pathCount(caveMap, partOne) << '\n'  // 4304
            << pathCount(caveMap, partTwo) << '\n'; // 118242
}
